(feat) Refactor secure configs stroing

petioptrv opened the pull request:
**Before submitting this PR, please make sure**:

- [X] Your code builds clean without any errors or warnings
- [X] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:
This PR refactors the way we store secure configs.

For strategies, the secure configs can be stored alongside
the rest of the strategy configs. They will be loaded with
the strategy configs and have the same lifetime.

For connectors, the login values / connector configs are
loaded on startup and are stored in the `Security` object
as previously. This allows for the export keys command
and several other parts of the code to continue working
as expected.

Connector configs are now stored in the same format as
strategy configs, and the secure values are no longer
separated in individual files.

All connectors have been refactored to this new approach to
configs storing.

The secure configs are no longer stored in the global
config map.

Finally, the PR introduces a new login step that checks for
existing configs and runs a migration script from the CLI.


**Tests performed by the developer**:
- Unit tests.
- Manually tested adding new connection.
- Manually tested loading existing connections.
- Manually tested creating new strategy with new and legacy configs.


**Tips for QA testing**:
- All exchanges' login information setup needs testing, but no need to test the function of the connectors themselves, simply to see if a connection can be created and validated.
- Strategy config creation (both new and legacy) needs re-testing as the path where those are stored has changed (no need to re-test amm thoroughly, simply if the config creation process works).
- I did not manage to manually test celo, so that will require testing as well.
- The export keys and balances commands should also be re-tested as they have been affected.

[ch21826]

gerald-hb commented:
Found several bugs in the commits.

1. Getting this long error message when you try to put the correct API key (given that the first attempt was incorrect)
![image](https://user-images.githubusercontent.com/84826167/166861726-a391136d-19a5-40ae-b145-d15a3f511dab.png)

Steps to reproduce
-installed featured branch
-connect an incorrect API (any exchange)
-when you attempt to reconnect the API, you will see this long error message

2. It accepts incorrect gate.io API credential
![image](https://user-images.githubusercontent.com/84826167/166861885-06746b3c-eb5d-4dd6-bb15-ee84a3e51454.png)

3.Keeps getting this error upon starting strategy (tested on pure market making and liquidity mining)
`2022-05-05 00:46:07,893 - 12935 - hummingbot.client.hummingbot_application - ERROR - Unknown error during initialization.
Traceback (most recent call last):
 File "/home/ubuntu/pr229/hummingbot/strategy/pure_market_making/start.py", line 89, in start
    self._initialize_markets(market_names)
  File "/home/ubuntu/pr229/hummingbot/client/hummingbot_application.py", line 277, in initializemarkets
    Security.update_config_map(global_config_map)`
    
Steps to reproduce
-install featured branch
-connect API keys
-create strategy (you can use default/basic config)
-start the strategy.
logs - 
[logs_binancetest.log](https://github.com/CoinAlpha/hummingbot/files/8628629/logs_binancetest.log)

Gateway Test
1.Getting an error when I try to create gateway
![image](https://user-images.githubusercontent.com/84826167/166862174-4c1f671f-9e17-4db9-9843-98399e4eee0c.png)


Celo Test
1.Unable to create an instance from the docker image that was created. I am getting this message
![image](https://user-images.githubusercontent.com/84826167/166862317-8362d0fd-09ec-40a7-a31b-424bc028db54.png)

I tried to copy manually the conf_migration_scripts.py but I keep getting an "invalid password" error
![image](https://user-images.githubusercontent.com/84826167/166862431-6bac9d8f-067d-434e-9e78-7e243c47ab1d.png)

2.Docker instance of celo-ultralight-node just keeps on restarting.
![image](https://user-images.githubusercontent.com/84826167/166862504-060b59a1-c326-4fba-995a-4926a684cbad.png)

PtrckM commented:
- typo on https://github.com/CoinAlpha/hummingbot/pull/229/files#diff-37174a189e5790bd11f8cd6a279d1692b747d2735c1859e1ca7baee82c5e6970R62 causing to fail when doing `./install` on macos

![image](https://user-images.githubusercontent.com/54516858/166865279-325c7709-c39d-427c-a19d-32c5941c73ca.png)

petioptrv commented:
@gerald-hb, @PtrckM, thank you for testing. I've addressed the flagged issues. I have also updated the config migration process. To test it, please use a folder containing old configs and delete any existing `.password_verification` file.

PtrckM commented:
> * typo on https://github.com/CoinAlpha/hummingbot/pull/229/files#diff-37174a189e5790bd11f8cd6a279d1692b747d2735c1859e1ca7baee82c5e6970R62 causing to fail when doing `./install` on macos
> 

Confirming the issue on macOS are now resolved ✅ 

---

- unable to launch client 
  - still getting the error even fresh install and new clones
  - both macos and cloud hit same issue

![image](https://user-images.githubusercontent.com/54516858/167046738-c363a273-7b40-45a7-bf47-e6176b5d0823.png)

petioptrv commented:
@PtrckM, sorry about that, I hadn't pushed an important change. I've added some additional fixes here too — if you try to overwrite an existing connection, but the new creds fail, then the old ones should now be restored.

PtrckM commented:
Confirmed fix - validates and accepts previous password are now fix ✅ 

---

- docker build - docker instance migration
  - has success message for migration 
    - unable to load previous keys / connect command shows api keys are not connected
    - import command does not able to find any strategy 
    - created empty folders `connectors` and `strategies`


https://user-images.githubusercontent.com/54516858/167232572-84ab45fe-f2d1-499e-ba36-364fcb79f4cc.mov


![image](https://user-images.githubusercontent.com/54516858/167232341-eb59171a-30b3-403d-a911-278bfc39eede.png)
- strategies and encrypted keys are visible

gerald-hb commented:
Previous reported bugs looks fixed. 

I found another unexpected behaviors.

1. Unable to auto-start a strategy. Getting this error message.
![image](https://user-images.githubusercontent.com/84826167/167234696-149e7112-4ff4-4f28-a50a-72798d399344.png)
![image](https://user-images.githubusercontent.com/84826167/167234713-32913059-a7bf-43d8-8adf-de28cdd2374f.png)
```2022-05-07 02:38:16,633 - 11390 - root - ERROR - No module named "hummingbot.strategy.{'strategy': <hummingbot"
Traceback (most recent call last):
  File "/home/ubuntu/hummingbot/pr229/hummingbot/client/config/config_helpers.py", line 451, in get_strategy_starter_file
    strategy_module = __import__(f"hummingbot.strategy.{strategy}.start",
ModuleNotFoundError: No module named "hummingbot.strategy.{'strategy': <hummingbot"
2022-05-07 02:38:16,634 - 11390 - hummingbot.core.utils.async_utils - ERROR - Unhandled error in background task: 
Traceback (most recent call last):
  File "/home/ubuntu/hummingbot/pr229/hummingbot/core/utils/async_utils.py", line 9, in safe_wrapper
    return await c
  File "/home/ubuntu/hummingbot/pr229/hummingbot/client/command/start_command.py", line 90, in start_check
    self._initialize_strategy(self.strategy_name)
  File "/home/ubuntu/hummingbot/pr229/hummingbot/client/command/start_command.py", line 216, in _initialize_strategy
    raise NotImplementedError
NotImplementedError
```
Steps to reproduce
-create a strategy
-exit the bot
-run this command "bin/hummingbot_quickstart.py -f >strategy config> -p <password>" (ex. bin/hummingbot_quickstart.py -f ascendextest.yml -p 11)
-when the client start, you will receive the error above.
[logs_ascendextest.log](https://github.com/CoinAlpha/hummingbot/files/8644298/logs_ascendextest.log)


2. Getting an error when running a balance command. It still display the balance but you get this error on the Log Pane
![image](https://user-images.githubusercontent.com/84826167/167234913-ae746ab1-c5f4-4850-bf5c-8305162d37b0.png)
```2022-05-07 02:43:55,105 - 11390 - hummingbot.core.utils.async_utils - ERROR - Unhandled error in background task: hasattr(): attribute name must be string
Traceback (most recent call last):
  File "/home/ubuntu/hummingbot/pr229/hummingbot/core/utils/async_utils.py", line 9, in safe_wrapper
    return await c
  File "/home/ubuntu/hummingbot/pr229/hummingbot/client/command/balance_command.py", line 121, in show_balances
    celo_address = CELO_KEYS.celo_address if hasattr("celo_address", CELO_KEYS) else None
TypeError: hasattr(): attribute name must be string
```

gerald-hb commented:
There is also one unusual behavior I notice. After a fresh clone/install, you will get this error message when you run the `gateway create` command
![image](https://user-images.githubusercontent.com/84826167/167236173-aae32f86-401f-4661-8606-935cf197c30f.png)

When you exit the bot and restart the client, `gateway create` will start to work
![image](https://user-images.githubusercontent.com/84826167/167236222-4854bb91-ffde-4ae9-a401-689dd50c8d64.png)

Steps taken
1.clone/install featured branch
2.run `gateway create` - You will get the error
3.exit and restart the bot
4.run `gateway create` again and it should work.

petioptrv commented:
@PtrckM @gerald-hb I've addressed the issues you flagged in the latest commit.

Deepali6997 commented:
got a chance to test the following in regards with PR229:-

- The balance is showing correctly :- AscendEX, Binance, Binance_perpetual, Binance_us, bybit_perpetual, Gate_io, Huobi, KuCoin, OKX, Probit
- When giving incorrect credentials we’re getting error for all the above mentioned exchanges
- Only correct API creds are getting accepted which is the ideal behavior
- Created a strategy successfully 
- Started the strategy and no error were spotted
- Export keys command is working as expected

Deepali6997 commented:
getting the following error while running auto-start command

<img width="1313" alt="Screenshot 2022-05-10 at 5 08 44 PM" src="https://user-images.githubusercontent.com/100112937/167620068-3c5170de-67d8-4f17-b982-930f1fec8453.png">

Steps to reproduce 
1. Clone, install and start the feature branch
2. Create any strategy using paper trade (used binance_paper_trade)
3. Exit the client 
4. Run the following command : bin/hummingbot_quickstart.py -f CONFIG_FILE_NAME -p CONFIG_PASSWORD
(note: make changes in CONF_FILE_NAME and CONFIG_PASSWORD according to the .yml file created and password created, before running this command)

expected:
strategy should start automatically 

actual: 
We're getting the error in CLI
Please refer above image

gerald-hb commented:
Auto start is now working :white_check_mark: 

1.installed featured branch:white_check_mark: 
2.connect several connectors and check balance:white_check_mark: 
3.create a strategy in paper trade:white_check_mark: 
4.stop - exit bot and tried to auto start strategy:white_check_mark: 
5.create strategy using different connectors:white_check_mark: 
6.stop command after a few minutes of the strategy running.:white_check_mark: 

Gateway/amm_arb test
1.Create a gateway container and add an Infura API key:white_check_mark: 
2.Connect gateway connector API keys:white_check_mark: 
3.Create an AMM arbitrage strategy using the gateway connector:white_check_mark: 
4.Start the strategy and stop after confirming it is checking for arbitrage:white_check_mark: 


Docker Image
Found another issue when creating a bot from a docker image. I am getting this error message
![image](https://user-images.githubusercontent.com/84826167/167744950-2de43947-bde7-4a64-b971-f908d8cef9c2.png)

Steps done
1. clone and installed featured branch 
2. build a docker image based from the cloned featured branch 
3. Run ./create.sh
4.After you setup password, it will return to the screen above

petioptrv commented:
@gerald-hb I've addressed the latest flagged issue.

gerald-hb commented:
Hi @petioptrv For some reason I still got the same error message. Both in Local and AWS
![image](https://user-images.githubusercontent.com/84826167/167990243-e5f67ae5-42bd-4ff6-a33c-1c7da4162dba.png)
![image](https://user-images.githubusercontent.com/84826167/167990760-f8aa162c-3e45-4b7c-95d7-e3c9595e697d.png)

Test performed
-clone and installed featured branch
-build a docker image based from the cloned featured branch
-Run ./create.sh and select the created pr image
-After you setup password, it will return to the screen above

petioptrv commented:
@CoinAlpha/qa-team a heads up for the commit: please make sure to update the `create.sh` as that's where the key change was made to address the latest bug

RC-13 commented:
Confirming that it no longer crashes when creating a docker instance.

gerald-hb reviewed (approved):
Docker image was successfully created by using the updated ./create.sh script. Celo testing is Blocked due to a known issue.

PtrckM commented:
- unable to migrate keys from kraken

https://user-images.githubusercontent.com/54516858/168196735-7c7b18fb-b8f5-454e-9d7c-97cb085d9614.mov

![image](https://user-images.githubusercontent.com/54516858/168196838-f82d4879-d911-4515-b8b1-354a217478ec.png)

petioptrv commented:
@PtrckM the last two commits should address this issue

RC-13 commented:
I no longer see the error mentioned above, however, there is an issue recognizing the tier level:
![image](https://user-images.githubusercontent.com/57189990/168307734-b59947cd-a174-43ed-a949-620876447bfe.png)
```
2022-05-13 14:33:29,585 - 1 - hummingbot.core.utils.async_utils - ERROR - Unhandled error in background task: '7B2263727970746F223A207B22636970686572223A20226165732D3132382D637472222C2022636970686572706172616D73223A207B226976223A20223665373236353833303834616365333962313831636631336264613161313265227D2C202263697068657274657874223A2022333237383164316235333931663263626238646135373137222C20226B6466223A202270626B646632222C20226B6466706172616D73223A207B2263223A20313030303030302C2022646B6C656E223A2033322C2022707266223A2022686D61632D736861323536222C202273616C74223A20223434376265393536653864393238626333366536343936313736613562663338227D2C20226D6163223A202237356338653865336665376536633466613139643730303464643236336339646565316435313330636630333636653938396136613861303738636431323736227D2C202276657273696F6E223A20337D' is not a valid KrakenAPITier
ValueError: '7B2263727970746F223A207B22636970686572223A20226165732D3132382D637472222C2022636970686572706172616D73223A207B226976223A20223665373236353833303834616365333962313831636631336264613161313265227D2C202263697068657274657874223A2022333237383164316235333931663263626238646135373137222C20226B6466223A202270626B646632222C20226B6466706172616D73223A207B2263223A20313030303030302C2022646B6C656E223A2033322C2022707266223A2022686D61632D736861323536222C202273616C74223A20223434376265393536653864393238626333366536343936313736613562663338227D2C20226D6163223A202237356338653865336665376536633466613139643730303464643236336339646565316435313330636630333636653938396136613861303738636431323736227D2C202276657273696F6E223A20337D' is not a valid KrakenAPITier

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/hummingbot/hummingbot/core/utils/async_utils.py", line 9, in safe_wrapper
    return await c
  File "/home/hummingbot/hummingbot/client/command/connect_command.py", line 102, in show_connections
    df, failed_msgs = await self.connection_df()
  File "/home/hummingbot/hummingbot/client/command/connect_command.py", line 117, in connection_df
    err_msgs = await asyncio.wait_for(
  File "/home/hummingbot/miniconda3/envs/secure-config/lib/python3.8/asyncio/tasks.py", line 483, in wait_for
    return fut.result()
  File "/home/hummingbot/hummingbot/user/user_balances.py", line 104, in update_exchanges
    results = await safe_gather(*tasks)
  File "/home/hummingbot/hummingbot/core/utils/async_utils.py", line 22, in safe_gather
    return await asyncio.gather(*args, **kwargs)
  File "/home/hummingbot/hummingbot/user/user_balances.py", line 79, in update_exchange_balance
    return await self.add_exchange(exchange_name, **api_keys)
  File "/home/hummingbot/hummingbot/user/user_balances.py", line 55, in add_exchange
    market = UserBalances.connect_market(exchange, **api_details)
  File "/home/hummingbot/hummingbot/user/user_balances.py", line 23, in connect_market
    connector = connector_class(**init_params)
  File "hummingbot/connector/exchange/kraken/kraken_exchange.pyx", line 121, in hummingbot.connector.exchange.kraken.kraken_exchange.KrakenExchange.__init__
    self._kraken_api_tier = KrakenAPITier(kraken_api_tier.upper())
  File "/home/hummingbot/miniconda3/envs/secure-config/lib/python3.8/enum.py", line 304, in __call__
    return cls.__new__(cls, value)
  File "/home/hummingbot/miniconda3/envs/secure-config/lib/python3.8/enum.py", line 595, in __new__
    raise exc
  File "/home/hummingbot/miniconda3/envs/secure-config/lib/python3.8/enum.py", line 579, in __new__
    result = cls._missing_(value)
  File "/home/hummingbot/miniconda3/envs/secure-config/lib/python3.8/enum.py", line 608, in _missing_
    raise ValueError("%r is not a valid %s" % (value, cls.__name__))
ValueError: '7B2263727970746F223A207B22636970686572223A20226165732D3132382D637472222C2022636970686572706172616D73223A207B226976223A20223665373236353833303834616365333962313831636631336264613161313265227D2C202263697068657274657874223A2022333237383164316235333931663263626238646135373137222C20226B6466223A202270626B646632222C20226B6466706172616D73223A207B2263223A20313030303030302C2022646B6C656E223A2033322C2022707266223A2022686D61632D736861323536222C202273616C74223A20223434376265393536653864393238626333366536343936313736613562663338227D2C20226D6163223A202237356338653865336665376536633466613139643730303464643236336339646565316435313330636630333636653938396136613861303738636431323736227D2C202276657273696F6E223A20337D' is not a valid KrakenAPITier
```
I endorse the issue to Patrick for re-checking.

Deepali6997 commented:
Can confirm the above flagged issue is no longer seen. 
No issue in recognising tier level.

PFB

<img width="1705" alt="Screenshot 2022-05-16 at 7 01 26 PM" src="https://user-images.githubusercontent.com/100112937/168604112-471d91d9-171d-48fb-9be2-fb7aad6533e0.png">

RC-13 reviewed (approved):
LGTM!

Test Summary:
   -  Test on source build
        * Clone and install the feature branch without any issue
        * Copy API keys and strategy files created from different branches and confirmed getting a migration message once the client is started
        * Run the connect command: Ok
        * Run the balance command: OK
        * Export keys: Ok
        * Export a strategy created from another instance and start: Ok
        * Create a new strategy and start: Ok
   - Test on docker build
        * Build a docker image based on the feature branch: Ok
        * Download the updated create script: Ok
        * Create a docker instance using the development branch: OK
        * Connect different exchanges: OK
        * Create a PMM strategy, Avellaneda strategy, and LM strategy: OK
        * Exit from the client and delete the docker image (based on the development branch): OK
        * Create a docker instance using the build image (feature branch) and name the docker instance with the same name used in the initial docker instance created.
        * Confirm migration image: Ok
        * Run connect command: Ok
        * Run the balance command: Ok
        * Import a strategy: Ok
        * Run the strategy: OK
        * Export keys: Ok
 - Skipped CELO_Arb due to a known issue

RC-13 commented:
Still LGTM for the latest commit.

aarmoa reviewed (changes requested):
I have added some comments.

I also have a general question: Is it a good idea to have .dockerignore and .gitignore files in several places? Is there a reason we can't add the ignore rules for those folders in the ignore files that are in the project root?

petioptrv commented:
> I also have a general question: Is it a good idea to have .dockerignore and .gitignore files in several places? Is there a reason we can't add the ignore rules for those folders in the ignore files that are in the project root?

For this, the reason behind adding the ignore files is to commit those folders to git, so that new installations come with the `strategies` and `connectors` sub-folders already available

petioptrv commented on hummingbot/client/config/conf_migration.py:
```diff
@@ -0,0 +1,137 @@
+                with open(key_path, 'r') as f:
+                    json_str = f.read()
+                value = binascii.hexlify(json_str.encode()).decode()
+                if not el.client_field_data.is_secure:
+                    value = Security.secrets_manager.decrypt_secret_value(el.attr, value)
```
I would expect the same, but if you take a look at the current Kraken utils, you'll see that the field `kraken_api_tier` is market as not secure, yet the secure configs management system dictates that it's still encrypted. That was a source of some of the bugs reported by QA.

petioptrv replied:
In other words, the current code encrypts all exchange config params regardless of marked as secure or not

aarmoa replied:
Oh, I see. That is weird, but I get it now.

petioptrv commented on hummingbot/client/ui/__init__.py:
```diff
@@ -1,10 +1,19 @@
+from hummingbot.client.config.security import Security
+from hummingbot.client.settings import CONF_DIR_PATH
+
+sys.path.insert(0, realpath(join(__file__, "../../../")))
```
Hmmm. i think this is done to resolve paths marked simply as `"/config"`. The new approach we're introducing to path specification (i.e. using `root_path`), makes this unnecessary, but I will keep it in case I've overlooked some legacy path definitions and I generally can't see it hurting to have as precaution.

I will change it to use `root_path` instead.

petioptrv commented:
@RC-13 nice catch! It was actually properly migrating only a couple of the strategies. It should now migrate all of them, but unfortunately they will all require testing as they're handled somewhat individually.

gerald-hb reviewed (approved):
LGTM!

Test on source build (Migration)
Clone and install the feature branch without any issue
Copy API keys and strategy files created from different branches and confirmed getting a migration message once the client is started
Run the connect command: Ok
Run the balance command: OK
Export keys: Ok
Export a strategy created from another instance and start: Ok
Create a new strategy and start: Ok

Test on docker build(Migration)
Build a docker image based on the feature branch: Ok
Download the updated create script: Ok
Create a docker instance using the development branch: OK
Connect different exchanges: OK
Create a PMM strategy, Avellaneda strategy, LM strategy, Cross exchange, perpetual market making, Arbitrage: OK
Exit from the client and delete the docker image (based on the development branch): OK
Create a docker instance using the build image (feature branch) and name the docker instance with the same name used in the initial docker instance created.
Confirm migration image: Ok
Run connect command: Ok
Run the balance command: Ok
Import a strategy: Ok
Run the strategy: OK
Export keys: Ok
Skipped CELO_Arb due to a known issue

aarmoa reviewed (approved):
Latest changes look good to me

aarmoa reviewed (approved):
Changes look good to me
